UnitTest/TestComponentCSharp_Tests.cs

```diff
@@ -1594,5 +1585,66 @@ public void TestNonAgileDelegateCall()
             observable.Add(3);
             Assert.Equal(6, observable.Observation);
         }
+
+        [ComImport]
+        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+        [Guid("EECDBF0E-BAE9-4CB6-A68E-9598E1CB57BB")]
+        internal interface IWindowNative
+        {
+            IntPtr WindowHandle { get; }
+        }
+
+        [ComImport]
+        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+        [Guid("3E68D4BD-7135-4D10-8018-9FB6D9F33FA1")]
+        internal interface IInitializeWithWindow
+        {
+            void Initialize(IntPtr hwnd);
+        }
+
+        [Fact]
+        unsafe public void TestComImports()
+        {
+            static Object MakeObject()
+            {
+                Assert.Equal(0, ComImports.NumObjects);
+                var obj = ComImports.MakeObject();
+                Assert.Equal(1, ComImports.NumObjects);
+                return obj;
+            }
+
+            static void TestObject() => MakeObject();
+
+            static (IInitializeWithWindow, IWindowNative) MakeImports() 
+            { 
+                var obj = MakeObject();
+                var initializeWithWindow = obj.As<IInitializeWithWindow>();
+                var windowNative = obj.As<IWindowNative>();
+                return (initializeWithWindow, windowNative);
+            }
+
+            static void TestImports()
+            {
+                var (initializeWithWindow, windowNative) = MakeImports();
+                
+                GC.Collect();
+                GC.WaitForPendingFinalizers();
+
+                var hwnd = new IntPtr(0x12345678);
+                initializeWithWindow.Initialize(hwnd);
+                Assert.Equal(windowNative.WindowHandle, hwnd);
+            }
+
+            TestObject();
+            GC.Collect();
+            GC.WaitForPendingFinalizers();
+            Assert.Equal(0, ComImports.NumObjects);
+
+            TestImports();
+            GC.Collect();
+            GC.WaitForPendingFinalizers();
+            // TODO: refcount leaks for ComImports
+            // Assert.Equal(0, ComImports.NumObjects);
```

Comment:
here's the leak - NumObjects should be 0 here

Reply:
A single `GC.Collect()` may not be enough. This kind of validation is incredibly hard when dealing with the built in system. I would try calling `GC.Collect()` 5 or more times and then `GC.WaitForPendingFinalizers()`.

Reply:
I'll experiment with that, but didn't think it was that nondeterministic as:
a) we have a number of unit tests for CCWs that only do a single collect
b) I have a baseline for just the IInspectable object that demos a single collect

Reply:
I believe Aaron is correct here. The lifetimes all look reasonable to me, so it looks like we may have to run multiple GC collects to ensure we get the objects collected.

Reply:
The lifetimes of the C#/WinRT generated types are more easily handled since there's no extra magic keeping things alive a little longer. The built-in system lacks the simplicity of the C#/WinRT system, so objects may have their lifetime slightly extended.

Reply:
>but didn't think it was that nondeterministic

It is incredibly nondeterministic. It is quite possible that the unit test framework is helping out a lot here, but I would be surprised if a small console would give you the desired reliability. The `GC.Collect` is only going to work if there is needed pressure. The fact that a few lines up a `GC.Collect` was done indicates to me that the system has already cleaned up so it is good. This also means `GC.WaitForPendingFinalizers()` is probably not doing anything because the queue is empty.

Reply:
Thanks guys - I had a missing Release